# CAPSTONE Project - Neighborhood Clustering

### Importing the dataset - IBM Watson

In [2]:

import types
import pandas as pd
from botocore.client import Config
import ibm_boto3




In [3]:
df_data_1 = pd.read_csv("Capstone_project.csv")
df_data_1.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

### The folloiwng is the shape of the dataset

In [4]:
df_data_1.shape

(288, 3)

### Examining how may boroughs have "Not Assigned" status

In [5]:
df_data_1.Borough.value_counts()

Not assigned        77
Etobicoke           45
North York          38
Downtown Toronto    37
Scarborough         37
Central Toronto     17
West Toronto        13
York                 9
East Toronto         7
East York            6
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64

#### In the Dataset, 77 entries of borroughs have "Not Assigned Status"

### Since the boroughs which have "Not assgined" status needs to be ignored, I need to drop the same from the dataset. For this, the indexes of such "Not assigned" status is found out. There should be total 77 such entries.

In [6]:
indexnames = df_data_1[df_data_1['Borough'] == "Not assigned"].index
indexnames

Int64Index([  0,   1,   9,  13,  20,  21,  30,  36,  37,  45,  46,  50,  51,
             52,  54,  55,  59,  60,  61,  73,  74,  75,  88,  89,  90, 104,
            105, 106, 120, 121, 136, 137, 148, 149, 155, 161, 162, 167, 175,
            181, 182, 188, 189, 190, 194, 195, 201, 202, 203, 204, 209, 210,
            223, 224, 237, 238, 241, 242, 247, 248, 253, 254, 258, 259, 260,
            261, 263, 264, 274, 275, 276, 277, 278, 279, 280, 281, 287],
           dtype='int64')

### We drop these entries where Boroughs have "Not assigned" status rom the dataset. Since original dataset had 288 rows and 77 of them were deleted, there would be 211 rows. We check the same

In [7]:
df_data_1.drop(indexnames, axis=0, inplace = True)
df_data_1.shape

(211, 3)

### We once again check if our exercise have been succesful.

In [8]:
df_data_1.Borough.value_counts()

Etobicoke           45
North York          38
Downtown Toronto    37
Scarborough         37
Central Toronto     17
West Toronto        13
York                 9
East Toronto         7
East York            6
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64

#### There are no "Not Assigned" Boroughs in the resulting dataset.

### Next, the number of unique Postcodes was found.

In [9]:
print("Number of Unique Postal codes is :", len(df_data_1.Postcode.unique()))

Number of Unique Postal codes is : 103


### Next, we check if there are any neighbourhood/s which have "Not Assigned" Status

In [10]:
df_data_1[df_data_1.Neighbourhood =='Not assigned']

Postcode       Borough Neighbourhood
8      M7A  Queen's Park  Not assigned

#### There is one entry where Neighbourhood has "Not assigned" status. We need to replace the Neighbourhood value with Borough name, which we do using the below code.

In [11]:
df_data_1.Neighbourhood.replace("Not assigned", "Queen's Park", inplace = True)

#### We check whether the same has been replaced.

In [12]:
df_data_1[df_data_1['Postcode'] == 'M7A']

Postcode       Borough Neighbourhood
8      M7A  Queen's Park  Queen's Park

### The neighbourhoods are joined using the below mentioned code using group by and lambda. It is saved in a different dataframe

In [13]:
df= pd.DataFrame(df_data_1.groupby("Postcode")['Neighbourhood'].apply(lambda tags: ','.join(tags)))
df

Neighbourhood
Postcode                                                   
M1B                                           Rouge,Malvern
M1C                    Highland Creek,Rouge Hill,Port Union
M1E                         Guildwood,Morningside,West Hill
M1G                                                  Woburn
M1H                                               Cedarbrae
M1J                                     Scarborough Village
M1K               East Birchmount Park,Ionview,Kennedy Park
M1L                           Clairlea,Golden Mile,Oakridge
M1M           Cliffcrest,Cliffside,Scarborough Village West
M1N                              Birch Cliff,Cliffside West
M1P       Dorset Park,Scarborough Town Centre,Wexford He...
M1R                                        Maryvale,Wexford
M1S                                               Agincourt
M1T                   Clarks Corners,Sullivan,Tam O'Shanter
M1V       Agincourt North,L'Amoreaux East,Milliken,Steel...
M1W                                         L'Amoreaux West
M1X                                             Upper Rouge
M2H                                       Hillcrest Village
M2J                              Fairview,Henry Farm,Oriole
M2K                                         Bayview Village
M2L                                 Silver Hills,York Mills
M2M                                  Newtonbrook,Willowdale
M2N                                        Willowdale South
M2P                                         York Mills West
M2R                                         Willowdale West
M3A                                               Parkwoods
M3B                                         Don Mills North
M3C                         Flemingdon Park,Don Mills South
M3H           Bathurst Manor,Downsview North,Wilson Heights
M3J                          Northwood Park,York University
...                                                     ...
M6C                                      Humewood-Cedarvale
M6E                                     Caledonia-Fairbanks
M6G                                                Christie
M6H                             Dovercourt Village,Dufferin
M6J                                 Little Portugal,Trinity
M6K              Brockton,Exhibition Place,Parkdale Village
M6L                        Downsview,North Park,Upwood Park
M6M             Del Ray,Keelesdale,Mount Dennis,Silverthorn
M6N                            The Junction North,Runnymede
M6P                            High Park,The Junction South
M6R                                   Parkdale,Roncesvalles
M6S                                       Runnymede,Swansea
M7A                                            Queen's Park
M7R                   Canada Post Gateway Processing Centre
M7Y       Business Reply Mail Processing Centre 969 Eastern
M8V              Humber Bay Shores,Mimico South,New Toronto
M8W                                   Alderwood,Long Branch
M8X             The Kingsway,Montgomery Road,Old Mill North
M8Y       Humber Bay,King's Mill Park,Kingsway Park Sout...
M8Z       Kingsway Park South West,Mimico NW,The Queensw...
M9A                                        Islington Avenue
M9B       Cloverdale,Islington,Martin Grove,Princess Gar...
M9C       Bloordale Gardens,Eringate,Markland Wood,Old B...
M9L                                           Humber Summit
M9M                                         Emery,Humberlea
M9N                                                  Weston
M9P                                               Westmount
M9R       Kingsview Village,Martin Grove Gardens,Richvie...
M9V       Albion Gardens,Beaumond Heights,Humbergate,Jam...
M9W                                               Northwest

[103 rows x 1 columns]

### Groupby statment results in indexing the Key. The same can be seen as below.

In [14]:
df.head(5)

Neighbourhood
Postcode                                      
M1B                              Rouge,Malvern
M1C       Highland Creek,Rouge Hill,Port Union
M1E            Guildwood,Morningside,West Hill
M1G                                     Woburn
M1H                                  Cedarbrae

### The indexing is removed below.

In [15]:
df.reset_index(inplace = True)
df.head(5)

Postcode                         Neighbourhood
0      M1B                         Rouge,Malvern
1      M1C  Highland Creek,Rouge Hill,Port Union
2      M1E       Guildwood,Morningside,West Hill
3      M1G                                Woburn
4      M1H                             Cedarbrae

### Next, we need to merge the 2 dataframes on Postcode for merging Neighbourhood. Hnece first, we drop the "Neighbourhood column from the original dataframe.

In [16]:
df_data_1.drop('Neighbourhood', axis= 1, inplace =True)
df_data_1.head(5)

Postcode           Borough
2      M3A        North York
3      M4A        North York
4      M5A  Downtown Toronto
5      M5A  Downtown Toronto
6      M6A        North York

### Merging the two dataframes.

In [17]:
df_final = pd.merge(df_data_1, df, on= 'Postcode')
df_final

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
..       ...               ...   
181      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
186      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
190      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
192      M5X  Downtown Toronto   
193      M8X         Etobicoke   
194      M8X         Etobicoke   
195      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
199      M8Y         Etobicoke   
200      M8Y         Etobicoke   
201      M8Y         Etobicoke   
202      M8Y         Etobicoke   
203      M8Y         Etobicoke   
204      M8Y         Etobicoke   
205      M8Y         Etobicoke   
206      M8Z         Etobicoke   
207      M8Z         Etobicoke   
208      M8Z         Etobicoke   
209      M8Z         Etobicoke   
210      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                             Harbourfront,Regent Park  
4                      Lawrence Heights,Lawrence Manor  
5                      Lawrence Heights,Lawrence Manor  
6                                         Queen's Park  
7                                     Islington Avenue  
8                                        Rouge,Malvern  
9                                        Rouge,Malvern  
10                                     Don Mills North  
11                      Woodbine Gardens,Parkview Hill  
12                      Woodbine Gardens,Parkview Hill  
13                             Ryerson,Garden District  
14                             Ryerson,Garden District  
15                                           Glencairn  
16   Cloverdale,Islington,Martin Grove,Princess Gar...  
17   Cloverdale,Islington,Martin Grove,Princess Gar...  
18   Cloverdale,Islington,Martin Grove,Princess Gar...  
19   Cloverdale,Islington,Martin Grove,Princess Gar...  
20   Cloverdale,Islington,Martin Grove,Princess Gar...  
21                Highland Creek,Rouge Hill,Port Union  
22                Highland Creek,Rouge Hill,Port Union  
23                Highland Creek,Rouge Hill,Port Union  
24                     Flemingdon Park,Don Mills South  
25                     Flemingdon Park,Don Mills South  
26                                    Woodbine Heights  
27                                      St. James Town  
28                                  Humewood-Cedarvale  
29   Bloordale Gardens,Eringate,Markland Wood,Old B...  
..                                                 ...  
181  Albion Gardens,Beaumond Heights,Humbergate,Jam...  
182           

### Sicne there were 211 rows in the original dataframe, the resulting merged dataframe will also contain 211 rows, as shown below. 

In [18]:
df_final.shape

(211, 3)

### However, this dataframe contains many duplicate values since the number of unqiue postcodes is only 103. We drop the duplicates ans check the final shape, which should contain 103 rows.

In [19]:
df_final.drop_duplicates(inplace = True)
df_final.shape

(103, 3)

### resetting the index

In [20]:
df_final.reset_index(inplace =True)
df_final

index Postcode           Borough  \
0        0      M3A        North York   
1        1      M4A        North York   
2        2      M5A  Downtown Toronto   
3        4      M6A        North York   
4        6      M7A      Queen's Park   
5        7      M9A         Etobicoke   
6        8      M1B       Scarborough   
7       10      M3B        North York   
8       11      M4B         East York   
9       13      M5B  Downtown Toronto   
10      15      M6B        North York   
11      16      M9B         Etobicoke   
12      21      M1C       Scarborough   
13      24      M3C        North York   
14      26      M4C         East York   
15      27      M5C  Downtown Toronto   
16      28      M6C              York   
17      29      M9C         Etobicoke   
18      33      M1E       Scarborough   
19      36      M4E      East Toronto   
20      37      M5E  Downtown Toronto   
21      38      M6E              York   
22      39      M1G       Scarborough   
23      40      M4G         East York   
24      41      M5G  Downtown Toronto   
25      42      M6G  Downtown Toronto   
26      43      M1H       Scarborough   
27      44      M2H        North York   
28      45      M3H        North York   
29      48      M4H         East York   
..     ...      ...               ...   
73     130      M4R   Central Toronto   
74     131      M5R   Central Toronto   
75     134      M6R      West Toronto   
76     136      M7R       Mississauga   
77     137      M9R         Etobicoke   
78     141      M1S       Scarborough   
79     142      M4S   Central Toronto   
80     143      M5S  Downtown Toronto   
81     145      M6S      West Toronto   
82     147      M1T       Scarborough   
83     150      M4T   Central Toronto   
84     152      M5T  Downtown Toronto   
85     155      M1V       Scarborough   
86     159      M4V   Central Toronto   
87     164      M5V  Downtown Toronto   
88     171      M8V         Etobicoke   
89     174      M9V         Etobicoke   
90     182      M1W       Scarborough   
91     183      M4W  Downtown Toronto   
92     184      M5W  Downtown Toronto   
93     185      M8W         Etobicoke   
94     187      M9W         Etobicoke   
95     188      M1X       Scarborough   
96     189      M4X  Downtown Toronto   
97     191      M5X  Downtown Toronto   
98     193      M8X         Etobicoke   
99     196      M4Y  Downtown Toronto   
100    197      M7Y      East Toronto   
101    198      M8Y         Etobicoke   
102    206      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                      Lawrence Heights,Lawrence Manor  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                        Rouge,Malvern  
7                                      Don Mills North  
8                       Woodbine Gardens,Parkview Hill  
9                              Ryerson,Garden District  
10                                           Glencairn  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  
12                Highland Creek,Rouge Hill,Port Union  
13                     Flemingdon Park,Don Mills South  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  
18                     Guildwood,Morningside,West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  C

### Removing the index column,

In [21]:
df_final.drop('index', axis=1, inplace = True)
df_final

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
..       ...               ...   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                      Lawrence Heights,Lawrence Manor  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                        Rouge,Malvern  
7                                      Don Mills North  
8                       Woodbine Gardens,Parkview Hill  
9                              Ryerson,Garden District  
10                                           Glencairn  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  
12                Highland Creek,Rouge Hill,Port Union  
13                     Flemingdon Park,Don Mills South  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  
18                     Guildwood,Morningside,West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28       Bathurst Manor,Downsview North,Wilson Heights  
29                                    Thorncliffe Park  
..                                                 ...  
73                                  North Toronto West  
74            

## The final shape of the dataframe:

In [22]:
df_final.shape

(103, 3)

# The final dataframe consists of 103 rows and 3 columns

###  